modules/core/ui/location/src/Plugin/Derivative/AddLocationAssetAction.php

```diff
@@ -0,0 +1,77 @@
+<?php
+
+namespace Drupal\farm_ui_location\Plugin\Derivative;
+
+use Drupal\Component\Plugin\Derivative\DeriverBase;
+use Drupal\Core\Entity\EntityTypeManagerInterface;
+use Drupal\Core\Plugin\Discovery\ContainerDeriverInterface;
+use Drupal\Core\StringTranslation\StringTranslationTrait;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+
+/**
+ * Defines actions to add location assets on locations page.
+ */
+class AddLocationAssetAction extends DeriverBase implements ContainerDeriverInterface {
+
+  use StringTranslationTrait;
+
+  /**
+   * The entity type manager service.
+   *
+   * @var \Drupal\Core\Entity\EntityTypeManagerInterface
+   */
+  protected $entityTypeManager;
+
+  /**
+   * Constructs a FarmActions object.
+   *
+   * @param \Drupal\Core\Entity\EntityTypeManagerInterface $entity_type_manager
+   *   The entity type manager service.
+   */
+  public function __construct(EntityTypeManagerInterface $entity_type_manager) {
+    $this->entityTypeManager = $entity_type_manager;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container, $base_plugin_id) {
+    return new static(
+      $container->get('entity_type.manager'),
+    );
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getDerivativeDefinitions($base_plugin_definition) {
+
+    // Get asset types that are locations by default.
+    $asset_storage = $this->entityTypeManager->getStorage('asset_type');
+    $asset_type_ids = $asset_storage->getQuery()
+      ->accessCheck(TRUE)
+      ->condition('status', TRUE)
+      ->exists('third_party_settings.farm_location')
+      ->condition('third_party_settings.farm_location.is_location', TRUE)
+      ->execute();
+
+    // Bail if there are no asset types.
+    if (empty($asset_type_ids)) {
+      return parent::getDerivativeDefinitions($base_plugin_definition);
+    }
+
+    // Add links to asset/add/type for each asset type.
+    $asset_label = $this->entityTypeManager->getDefinition('asset')->getLabel();
+    foreach ($asset_storage->loadMultiple($asset_type_ids) as $asset_type_id => $asset_type) {
+      $this->derivatives[$asset_type_id] = $base_plugin_definition;
+      $this->derivatives[$asset_type_id]['title'] = $this->t('Add @bundle @entity_type', ['@bundle' => $asset_type->label(), '@entity_type' => $asset_label]);
+      $this->derivatives[$asset_type_id]['route_name'] = 'entity.asset.add_form';
+      $this->derivatives[$asset_type_id]['route_parameters'] = ['asset_type' => $asset_type_id];
+      $this->derivatives[$asset_type_id]['appears_on'][] = 'farm.locations';
+      $this->derivatives[$asset_type_id]['appears_on'][] = 'farm.asset.locations';
```

Comment:
Right now this will show the action link on `/asset/%/locations`, which is where we display the sub-hierarchy of locations under the current asset (if that asset is a location itself). But the action button will just take you to a blank asset creation form without pre-populating anything.

I think as a user, if I saw that button in that context, my expectation would be that it would be for adding more children locations, and so it would pre-populate the `parent` field in the asset creation form.

This would be similar to the way in which clicking "Add Activity log" on an asset automatically references that asset in the log edit form.

This would be nice, but I don't think we need to worry about it as part of this first-step PR. I think we should just omit this line for now and NOT show the button on `farm.asset.locations`. We can consider adding logic to pre-populate the `parent` field in the future.

Reply:
Agreed :+1: